### PR TITLE
Game logic

### PR DESCRIPTION
**Pull request by elasfarc**
In this milestone we did the following:

- create classes for Player and Board -> initialize objects when game starts

-Board has serval methods to check at anytime the Board's object status and alter it (check_input, wrong_input?, already_taken?, modify() a cell in matrix , game_finish? etc..)

- Player class has serval methods to check for the current_player(), switch_player() and whos_turn?()

- main is responsible for the game-logic (how the user interact with the game and how the game responds to the use's input )

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -0,0 +1,111 @@
+class Board
+  attr_accessor :board
```
Comment (kobiyoyo) [inline lib/board.rb:2]: - [ ] You should use  attr_readers or attr_writers. Only using attr_accessors it allows writing that property from outside the class. That variable can be set to anything. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/attr_accessor.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

File: lib/player.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# rubocop:disable Style/ClassVars
+class Player
+  @@players_arry = []
+  @@current_player = true
+  attr_accessor :name
```
Comment (kobiyoyo) [inline lib/player.rb:5]: - [ ] You should use  attr_readers or attr_writers. Only using attr_accessors it allows writing that property from outside the class. That variable can be set to anything. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/attr_accessor.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team    
#### Great Job so far, everything works fine, but there are a few more things required. 
### General
- [ ] All game-logic related resources (classes and/or modules) are kept in the lib directory.
- [x] I get an unending request whenever I want to play again 
- [ ] Lets add validation for empty inputs and integers.
- [ ]  Check comments below
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

File: lib/board.rb [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  end
+
+  def diagonal_check(_ary, letter)
+    ((@board[0][0] == letter) && (@board[1][1] == letter) && (@board[2][2] == letter)) ||
+      ((@board[0][2] == letter) && (@board[1][1] == letter) && (@board[2][0] == letter))
+  end
+end
+
+# board1 = Board.new
+# puts board1.print_board
```
Comment (euqueme) [inline lib/board.rb:111]: - [ ] remove these two lines of commented code

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request reviewed by Maria Eugenia Queme P. (TSE)

### Hello @elasfarc and @karmaester :balloon: :lollipop: 

### Thank you for doing the changes requested by the previous reviewer :medal_sports: :medal_sports: your game works well :ok_hand: and your code is impeccable I particularly like how complete your instructions are and that you request for the player's names I think this makes your game very human friendly and stand out from the rest :video_game: :star: there's only a couple of things left to work on :construction: :construction_worker: :construction_worker: 

#### Behavior

- [ ] The game doesn't restart whenever you want to play again, instead, it asks for the second player's Name and after that asks if you want to play again so it's stuck in a loop, let's fix it or since it wasn't requested remove it from the game.

image

#### General 

- [ ] Comments: It’s better to delete **commented piece of codes**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

> I've left comments in your code where I've found commented code that needs to be deleted

### That's it, team! :raised_hands: 

### Please :pray: do the changes :technologist: :technologist: and submit back for a code review when you are done :memo: 

### Happy coding! :rocket: 

Any questions? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

**Review by kitpao — APPROVED** [review]
# 👋 Hello @elasfarc and @karmaester 

Great job implementing all required changes 🌟 Your project is **approved**. 
📝 In case you want to learn more, here I leave some optionals:

**_Note:_** All the **[OPTIONAL]** are **not required** in order to approve your project but are personal suggestions to improve your coding experience

- [ ] **OPTIONAL** Your game accepts 2 players with the same name, maybe you could add that validation

- [ ] **OPTIONAL** I see you disabled Rubocop's Style/ClassVars. After inspecting I see that you are not really using the instance variables, you are using Class Variables. You are modifying the class with your methods instead of modifying only the instances.
Please take a look at the difference in code and the 3 comments inside for how to use instance variables:

**Doc: lib/player.rb**
```ruby
class Player
  
  @current_player = true
  attr_reader :name

  def initialize(name = 'unknown_player')
    @name = name
    @players_arry = []         # <---------- 1) instance variables declared here, it doesn't makes sense to put it as a player variable, it would make more sense if you had a Class Game, but this works too
    @current_player = true
    @players_arry.push(@name)
  end

  def switch_player             # <---------- 2) all methods should be declared without the self for instance variables. Again, this would make more sense for a Class Game
    @current_player = !@current_player
  end

  def current_player
    @current_player
  end

  def whos_turn?
    if current_player
      "okay #{@players_arry[0]} it's ur turn "
    else
      "okay #{@players_arry[1]} it's ur turn "
    end
  end

  def play_again?(arg)
    arg == 'y'
  end

  def valid_name?(player_name)
    player_name != '' && player_name != ' ' && player_name.to_i.zero? &&
      ((player_name =~ /0/) != 0)
  end

  def reset
    @players_arry = []
    @current_player = true
  end
end
```

**Doc: bin/main**
```ruby
#!/usr/bin/env ruby

require_relative '../lib/player.rb'
require_relative '../lib/board.rb'

outputs = {
  ask_input: 'Write a letter from A to C and a number from 1 to 3',
  invalid_name: 'My master name can not be empty or just a number, let\'s try again! ',
  wrong: 'wrong input',
  taken: 'already taken',
  draw: 'it\'s a draw',
  winner: 'Congratulations you are the winner!',
  play_again: 'Do you want to keep playing? (y) to play again/ any key to exit'
}
input = ''
keep_going = true

def welcome
  puts 'Welcome to the Tic Tac Toe game!'
  puts 'do you want to proceed? (y/n)'
  user_answer = gets.chomp.downcase
  if user_answer.match(/y/)
    true
  elsif user_answer.match(/n/)
    puts 'bye'
    false
  else
    puts 'wrong answer'
    welcome
  end
end

def instructions
  puts "Instructions:\n
   1)The game is played on a grid that's 3 squares by 3 squares.\n
   2)Write a letter from A to C\n
    and a number from 1 to 3 \n
    to select the row and the column respectively that you want to play on.\n
   3)Player 1 is X, player 2 is O.\n
   4)Once a player gets 3 marks in a row\n
    (up, down, across, or diagonally) is the winner.\n
   5)If all 9 squares are full, the game is over \n \n"
end

# game starts
while keep_going
  valid_name = false
  game_on = true
  break unless welcome

  instructions
  until valid_name
    puts "Okay master! May I have first player's name?"
    player_name = gets.chomp
    p1 = Player.new(player_name)
    valid_name = p1.valid_name?(p1.name)    # <---------- 3) You call the methods on the instance object, not on the whole class
    unless valid_name
      puts outputs[:invalid_name]
      next
    end
    puts "cool! #{p1.name}"
  end
  valid_name = false
  until valid_name
    puts "what about second player's name?"
    player_name = gets.chomp
    p2 = Player.new(player_name)
    valid_name = p2.valid_name?(p2.name)
    unless valid_name
      puts outputs[:invalid_name]
      next
    end
    puts "cool! #{p2.name}, let's start to play "
  end

  board = Board.new
  puts board.p_board
  while game_on

    puts p1.whos_turn?
    loop do
      puts outputs[:ask_input]
      input = gets.chomp.upcase

      input_status = board.check_input(input)

      unless input_status == 'accepted'
        puts outputs[input_status.to_sym]
        next
      end

      board.modify(p1.current_player, input)
      puts board.p_board

      break
    end

    game_status = board.game_finish?(p1.current_player, input)

    if game_status
      puts outputs[game_status.to_sym]
      game_on = !game_on
    end
    p1.switch_player
  end
  puts outputs[:play_again]
  keep_going = p1.play_again?(gets.chomp.downcase)
  p1.reset if keep_going
end
```

### 👏 Your project has been approved 👏
### 🚀 You may merge to master and submit the project completition form 🚀

___
##### 👉 If you have any doubts you can contact me in slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate if can leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😛 

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Comment by elasfarc** [comment]
> * [ ]  **OPTIONAL** Your game accepts 2 players with the same name, maybe you could add that validation
> * [ ]  **OPTIONAL** I see you disabled Rubocop's Style/ClassVars. After inspecting I see that you are not really using the instance variables, you are using Class Variables. You are modifying the class with your methods instead of modifying only the instances.

Thank you for your review
we will implement your suggestions on milestone #4 
the only reason we chose to make [players_arry] and [current_player] as class variables instead of instance variables is because it does not make sense to say something like 
`  player1 = Player.new(playerName) 
`then
` player1.current_player 
`or 
`player_one.players_arry
` but it makes sense if Player class is a Game class

**Pull request merged**